zcash_primitives/src/merkle_tree.rs

```diff
@@ -117,7 +121,35 @@ impl<Node> CommitmentTree<Node> {
         }
     }
 
-    pub fn to_frontier(&self) -> bridgetree::Frontier<Node, SAPLING_COMMITMENT_TREE_DEPTH_U8>
+    pub fn from_frontier<const DEPTH: u8>(frontier: &bridgetree::Frontier<Node, DEPTH>) -> Self
+    where
+        Node: Clone,
+    {
+        frontier.value().map_or_else(Self::empty, |f| {
+            let (left, right) = match f.leaf() {
+                Leaf::Left(v) => (Some(v.clone()), None),
+                Leaf::Right(l, r) => (Some(l.clone()), Some(r.clone())),
+            };
+            let mut ommers_iter = f.ommers().iter().cloned();
+            let upos: usize = f.position().into();
+            Self {
+                left,
+                right,
+                parents: (1..DEPTH)
```

Comment:
I was concerned there might be an off-by-one here, but I now believe this to be correct. `parents.len()` is at most `depth - 1`, because the final layer in the depth is stored in `left / right`. Starting from `1` is correct here because the first bit of `upos` corresponds to `left / right`.